### PR TITLE
Move to sideway/joi

### LINKED ISSUE #163: Move to new joi package
I saw that the package [`@hapi/joi`][1] is now deprecated, in favour of the new [`joi`][2] package. Is a move to the new packaged planned or already in the work?

[1]: https://www.npmjs.com/package/@hapi/joi
[2]: https://www.npmjs.com/package/joi

### PR DESCRIPTION
**Pull request by bl4ckb0ne**
Fixes #163 

v16.0.0 breaking changes https://github.com/sideway/joi/issues/2037
v17.0.0 breaking changes https://github.com/sideway/joi/issues/2262

Note on the migration from @hapi/joi to sideway/joi https://github.com/sideway/joi/issues/2411

### CODE REVIEW

**Comment by bl4ckb0ne** [comment]
@laurence-myers CI fails because of node 8.10, should I bump it or should I fix it?

I am also unable to reproduce locally (even when running on the `circleci/node:8.10` image). The image is also not listed anymore in [circleci supported images][1].

[1]: https://circleci.com/docs/2.0/circleci-images/#nodejs

**Comment by laurence-myers** [comment]
Thanks for tackling this! Great work so far. 😁

Yes, let's bump to Node 10. (Node 8 reached EOL in 2019)

**Comment by bl4ckb0ne** [comment]
Alright looks good on the CI side, I'll let you review it.

File: yarn.lock
Comment (codeandcats) [inline yarn.lock]: Hmmm, I'm surprised to see v15 here. You might need to regenerate the `yarn.lock` file.

Try running the following:

```sh
rm -rf ./**/node_modules
yarn
rm ./yarn.lock
yarn
```

File: test/unit/validation.test.ts
Comment (codeandcats) [inline test/unit/validation.test.ts]: Instead of removing this check, we should probably check that the schema's `validate` method was called as expected.

File: tsconfig.json
Comment (codeandcats) [inline tsconfig.json]: This is covered by the [`strict`](https://www.typescriptlang.org/tsconfig#strict) option.

File: test/unit/testUtil.ts
Comment (codeandcats) [inline test/unit/testUtil.ts]: Hmm, I'm not sure if checking they both have the same keys is really good enough. Do we know why the previous check now fails? We might be able to simply ignore the specific properties that might be causing the issue?

**Review by codeandcats — CHANGES_REQUESTED** [review]
Wow, thank you so much for your contribution @bl4ckb0ne, it's seriously much appreciated. 🏆 🍻 

Unfortunately I don't have much time to review your PR this week. I've had a quick scan over it for now and provided some early feedback, but I'd like to do a more comprehensive review on the weekend when I'll have some free time. In the mean time please have a read over my comments and address/reply at your leisure. 😃

**Comment by codeandcats** [comment]
One more thing. Whenever we've changed the API of Joiful we need to include the word `BREAKING CHANGE` in the commit body with a short list of breaking changes. This is because we use commit conventions for knowing how to bump versions. Have a look at the examples [here](https://www.conventionalcommits.org/en/v1.0.0/) for more info.

File: yarn.lock
Comment (bl4ckb0ne) [reply yarn.lock]: I opened #167 because of the output from it. I'll fix the issue once this patch is merged.

Comment (bl4ckb0ne) [reply yarn.lock]: The commands you gave me updated everything not up to date, do you want that in that patch?

Comment (bl4ckb0ne) [reply yarn.lock]: Nevermind, fixed it with `yarn upgrade joi`

File: test/unit/validation.test.ts
Comment (bl4ckb0ne) [reply test/unit/validation.test.ts]: Done!

File: tsconfig.json
Comment (bl4ckb0ne) [reply tsconfig.json]: Done!

File: test/unit/testUtil.ts
Comment (bl4ckb0ne) [reply test/unit/testUtil.ts]: Ah yes, this one is a bit tricky. Keeping the original check gives the following error

```
Expected: [object omitted for brievety]
Received: serializes to the same string
```

I also tried to use `toMatchObject` and it gives the following error

```
expect(received).toMatchObject(expected)

    - Expected
    + Received

    @@ -68,11 +68,10 @@
          "untrace": [Function anonymous],
          "valid": [Function anonymous],
          "version": "17.3.0",
          "when": [Function anonymous],
          "x": [Function expression],
    -     Symbol(@hapi/lab/coverage/initialize): [Function trace],
        },
        "$_super": Object {
          "default": [Function bound default],
        },
        "$_temp": Object {
    @@ -157,11 +156,10 @@
                  "untrace": [Function anonymous],
                  "valid": [Function anonymous],
                  "version": "17.3.0",
                  "when": [Function anonymous],
                  "x": [Function expression],
    -             Symbol(@hapi/lab/coverage/initialize): [Function trace],
                },
                "$_super": Object {},
                "$_temp": Object {
                  "ruleset": false,
                  "whens": Object {},
```

Objects are 'almost' equals, but I couldn't find out where this tracing symbol came from. So I went with comparing the keys instead.

**Comment by bl4ckb0ne** [comment]
@codeandcats do you want me to put the joiful changes in the commit body or a link to the joi release note?

File: src/decorators/any.ts
Comment (codeandcats) [inline src/decorators/any.ts]: Very minor thing but could we revert this change please?

You should be able to call like this:
- `.tag('auth')`
- `.tag('auth', 'identity')`
- `.tag(['auth', 'identity'])`

But not like this:
- `.tag(['auth'], 'identity')`

Indirectly related FYI, I believe @laurence-myers's intention with using `.tag(tag: string, ...tags: string[])` instead of simply `.tag(...tags: string[])` is so at least one tag is required when calling the function with variadic parameters.

**Review by codeandcats — CHANGES_REQUESTED** [review]
Simon,

I'm so sorry for the late re-review of this fantastic Joiful contribution. Again I want to say we really appreciate your effort!

I've [created a PR](https://github.com/bl4ckb0ne/joiful/pull/1) to your branch that just addresses the weak checking that `assertClassSchemaEquals` function was doing. I think once you've got this in and the other minor thing I mentioned in this review, and reword your commits to include `BREAKING CHANGE: blah...` I think we're looking pretty golden. I would prefer a short description of the breaking change but I think a link to the specific release notes breaking changes for Joi might be fine too - @laurence-myers do you have an opinion here?

@laurence-myers I would love it though if you could cast your keen eyes across this too before we merge. Particularly across the new `link` constraint which I admittedly did not spend much time on.

Thanks

File: src/decorators/any.ts
Comment (bl4ckb0ne) [reply src/decorators/any.ts]: Done!

**Comment by bl4ckb0ne** [comment]
I squashed the commits in 3 distinctive commits (migration to joi 16, migration to joi 17 and the ci image bump). I also added the `BREAKING CHANGES` label with a link to the related release notes.

Let me know if there's anything else I can do.

File: src/decorators/any.ts
Comment (codeandcats) [reply src/decorators/any.ts]: Looks like this didn't get reverted

**Review by codeandcats — CHANGES_REQUESTED** [review]
Looks good @bl4ckb0ne. Just found that one thing re `tag` that didn't get reverted by the looks of it. 👌

**Comment by bl4ckb0ne** [comment]
Woopsy, i just removed it

File: src/decorators/any.ts
Comment (codeandcats) [inline src/decorators/any.ts]: This might sound like a nit but I think the original signature with the overload is still the most accurate. Can we please leave it as:

```typescript
    tags(tag: string, ...tags: string[]): this;	    tag(tag: string | string[], ...tags: string[]): this;
    tags(tags: string | string[]): this;
```

or

```typescript
    tags(tag: string, ...tags: string[]): this;	    tag(tag: string | string[], ...tags: string[]): this;
    tags(tags: string[]): this;
```

**Review by codeandcats — CHANGES_REQUESTED** [review]
Hi @bl4ckb0ne, sorry to be pedantic but can we revert the `tag` change back to its original signature please? After this all looks good on my side. Once this change is in and @laurence-myers is happy, I'm cool with merging this and cutting a new release. 😃

File: src/decorators/any.ts
Comment (bl4ckb0ne) [reply src/decorators/any.ts]: My bad, it's fixed.

**Review by codeandcats — APPROVED** [review]
Okay this is looking good, there's just one remaining issue I've noticed. The commit containing breaking changes needs to have "BREAKING CHANGE:" in the commit body but the current wording is slightly off so the release process does not bump the package version correctly.

In order to skip another feedback cycle I'm going to merge the PR then amend the commit description in master.

**Comment by codeandcats** [comment]
Okay, version joiful v3.0.0 is now published! Please raise an issue if you spot any problems.

Thank you again @bl4ckb0ne for your wonderful contributions and patience. You're a champion. 💯 🏆

**Pull request merged**
